specification/dartLangSpec.tex

```diff
@@ -11677,113 +12041,209 @@ \subsection{Identifier Reference}
+% to the static type of $p$. So we do not mention that type here.
+\commentary{%
+No such type is needed, because every construct where an import prefix $p$ is
+used and followed by \lit{.} is specified in such a way that the type
+of $p$ is not used.%
+}
+\EndCase
 
 \LMHash{}%
-The static type of $e$ is determined as follows:
+\Case{Lexical lookup yields a member signature}
+In this case the lexical lookup
+(\ref{lexicalLookup})
+for \id{} yields a member signature $s$.
+%
+In this situation $e$ is treated as
+(\ref{notation})
+\code{\THIS.\id}.
+
+\commentary{%
+This occurs when no declaration named \id{} is in scope,
+but $e$ occurs in the body of an instance member declaration,
+and the interface of the enclosing class has a member named \id.
+Both the static analysis and evaluation proceeds with
+\code{\THIS.\id},
+so there is no need to further specify the treatment of $e$.%
+}
+\EndCase
+
+\LMHash{}%
+Evaluation of an identifier expression $e$ of the form \id{}
+proceeds as follows:
+
+\LMHash{}%
+\Case{Lexical lookup yields a declaration}
+In this case the lexical lookup
+(\ref{lexicalLookup})
+for \id{} yields a declaration $D$.
+The evaluation of $e$ proceeds as follows:
 
 \begin{itemize}
-\item If $d$ is a class, type alias or type parameter the static type of $e$ is \code{Type}.
-\item If $d$ is a local variable (\commentary{which can be a formal parameter})
-  the static type of $e$ is the type of the variable \id,
-  unless \id{} is known to have some type $T$,
-  in which case the static type of $e$ is $T$,
-  provided that $T$ is a subtype of any other type $S$ such that $v$ is known to have type $S$.
-\item If $d$ is a static method, top-level function or local function the static type of $e$ is the function type defined by $d$.
-\item If $d$ is the declaration of a class variable, static getter or static setter declared in class $C$,
-  the static type of $e$ is the static type of the getter invocation (\ref{propertyExtraction}) \code{$C$.\id}.
-\item If $d$ is the declaration of a library variable, top-level getter or top-level setter,
-  the static type of $e$ is the static type of the top level getter invocation \id.
-\item Otherwise, if $e$ occurs inside a top level or static function (be it function, method, getter, or setter) or variable initializer,
-  the static type of $e$ is \DYNAMIC{}.
-\item Otherwise, the static type of $e$ is the type of the property extraction (\ref{propertyExtraction}) \code{\THIS.\id}.
+\item
+  If $D$ is a class, mixin, or type alias,
+  the value of $e$ is an object implementing the class \code{Type}
+  which reifies the corresponding type.
+\item
+  If $D$ is a type parameter $X$ then the value of $e$ is
+  the value of the actual type argument corresponding to $X$
+  that was passed to the generative constructor that created
+  the current binding of \THIS{}.
+\item
+  If $D$ is the declaration of
+  a library variable, top-level getter, or top-level setter,
```

Comment:
It can't be a setter, right? (Or variable).

Reply:
I already fixed the setter issue (don't know how it pops up here now).

But the variable issue is a little bit messy. We do have implicitly induced setters and getters, but we do not consistently say that those are the ones that we obtain from lookups. So a lookup may yield a variable declaration. So it would be a substantially bigger change to say consistently that lookups can never yield a non-local variable. Please create a separate issue if you wish to push that change.